### ISSUE TITLE
Implementation of the Expiration extension

### ISSUE THREAD

**Issue by Acconut** [tus/tusd#435]
tusd does currently not support the [Expiration](https://tus.io/protocols/resumable-upload.html#expiration) extension and people have to rely on custom scripts or provider-specific approaches to clean up old uploads. This issue is intended to serve as a discussion place for a possible implementation of this extension. In the end, tusd should be able to do two things:
- Provide the expiration time of an upload in the HEAD response to clients, and
- Be able to cleanup (finished and unfinished) uploads after a given time.

Sending the expiration time to the client should not be a though nut to crack. However, when looking at the actual clean up work, it could be hairy to find a solution for the various storage options that tusd supports:
- For the file storage, it is probably the best approach to iterate through all uploads on disk and determine which upload needs to be cleaned up. This could be implemented as an additional binary (maybe named `tusd-cleaner`) which the user must configure to run regularly as a cron job.
- When looking at the S3 and GCS storage, the cloud providers also offer some expiration mechanism built into their platform. For example, AWS S3 allows to delete files after a certain time. However, this mechanism has a low granularity since the expiration can only be configured in days. This approach is not helpful if someone wants to delete uploads after two hours. For these cloud providers we should consider providing two mechanisms: Either let the expiration be handled by the cloud provider itself or run the `tusd-cleaner` regularly which will iterate through your cloud bucket and find old uploads.

One might ask, we do we need two binaries (`tusd` for accepting uploads and `tusd-cleaner` for clean up old uploads)? Why can't we integrate the cleaning into tusd directly? Well, if you have multiple tusd servers running in parallel for redundancy and load balancing, you don't want every one of these tusd instances to scan through your upload directory at the same time. It would be better to let this be handled by one server with a cron job.

On the implementation side we would need following:
- A new directory for our `tusd-cleaner` binary in https://github.com/tus/tusd/tree/master/cmd
- A new datastore interfacer for iterating through existing uploads in https://github.com/tus/tusd/blob/master/pkg/handler/datastore.go
- Implementations for the new datastore interface in our storages (maybe try filestore first and later s3store and gcsstore)
- A new method for tusd to retrieve the expiration time of an uploads and include in the response to the client

These are just some ideas from me. I am open to hearing other thoughts and approaches!

**Comment by DravitLochan** [comment]
@Acconut Is this still open? I can pitch in for the design, discussion and implementation.

**Comment by Acconut** [comment]
@DravitLochan Yes, this task is still open. Any help is appreciated!

**Comment by DravitLochan** [comment]
@Acconut Shouldn't the overall file expiry and expiry of partial uploads be two different features? IMO TUS being a protocol (or a pseudo protocol), the overall file expiry should be out of the scope of TUS. That's something the applications should handle using a cron. Thoughts?

**Comment by alter123** [comment]
@DravitLochan I think your point of keeping different expiry for completed and partial uploads can be finely grained to configuration which allows user to define

- An overall expiry (based on first creation time of file)
- Different expiry for completed & partial uploads

One another opinion I would like to add is to clean uploads based on max available system disk (e.g. If the disk consumed is >90% then it should automatically delete the older uploads/start rejecting the new uploads).

**Comment by DravitLochan** [comment]
> One another opinion I would like to add is to clean uploads based on max available system disk (e.g. If the disk consumed is >90% then it should automatically delete the older uploads/start rejecting the new uploads).

Well, I am not sure of this as a feature for resumable-uploads. Also as a backend dev, to start deleting the older uploads on disk hitting 90% seems like an unacceptable thing to me. I might start to lose some critical files and hence my end user will start to churn. 

If you want to solve the above problem, I'd say the best way would be to have alerts in place for disk hitting 80, 85, 90 and 95%. In-fact if you run out of space, let the application handle it on it's end "due to high traffic, we are not allowing any new file uploads. We will be back up in no time".

**Comment by alter123** [comment]
@DravitLochan Your point makes sense, and rejecting the uploads sounds like correct option, but not sure if we need implementation for this at the moment, @Acconut can clarify.

Implementation wise I think storing expiry in [FileInfo](https://github.com/tus/tusd/blob/master/pkg/handler/datastore.go#L10) struct makes sense, since every FileSystem can have different abstraction for the same.

**Comment by DravitLochan** [comment]
@Acconut what do you think?

**Comment by Acconut** [comment]
> TUS being a protocol (or a pseudo protocol), the overall file expiry should be out of the scope of TUS.

Yes, tus should not be concerned about the expiry of files after the upload is completed, but tusd can definitely be concerned about this. There are multiple areas where tusd offers more features than the tus protocol specifies, just because those are handy and fit well into tusd. So, I see not problem when tusd offers expiration of uploads and files in one convenient package. Does that make sense? :)

> One another opinion I would like to add is to clean uploads based on max available system disk

Interesting idea, but I think that is outside of tusd's scope and is very application specific. This does not seem like a good fit for tusd, as @DravitLochan already mentioned.

> rejecting the uploads sounds like correct option, but not sure if we need implementation for this at the moment,

You can already check the free space of the disk in the pre-create hook and reject uploads if the disk is too full. There is not additional implementation needed inside tusd.

> Also I'm not sure how we can parse common flags for config etc across the binaries, since most of them will be same in `tusd` & `tusd-cleaner`

That should not be a big problem. The parsing and/or validation of flags can be extracted into a shared module, which is imported by both programs (if that makes sense in the case).

**Comment by Gambitier** [comment]
@Acconut 

I am currently using the `CreatedUploads` event to push tasks to a queue for cleaning up files. This way, I don't have to worry about which instance of tus created the file. So far, this approach is working well for us. Do you have any feedback or comments on this approach?

```go
for event := range handler.CreatedUploads {
		log.Printf("Upload %s created\n", event.Upload.ID)
		uploadedFilePath := hook.Upload.Storage["Path"]
		
		tusFiles := []string{
			uploadedFilePath,
			fmt.Sprintf("%v.info", uploadedFilePath),
		}
		
               // Add file cleanup task for tusFiles to message queue to be executed at the time of expiry
}
```

**Comment by Acconut** [comment]
That should be fine as long as the delay is large enough to make sure uploads are either finished or abandoned when the clean up occurs.

**Comment by rglKali** [comment]
Hey everyone, can I pick this task? I have a few thoughts on implementation. 

I think that a separate cron/daemon reaper will be a bit outside of the tusd responsibilities, as it requires of bookkeeping all the uploads, periodically checking the expired uploads, and cleaning them up. This will be a major change to how the Data Sources are implemented now, as there's no list-like methods, only id-based access. What we can do instead, is extend the Upload API with the SetExpiration/GetExpiration methods, allowing to adjust it manually via the hooks, and then just reject GET/PATCH/DELETE on the expired resources with the immediate termination of the upload -> cleaning up the resource. In addition, if the Data Source supports expiration, we can also add the PreExpiry callback and PostExpiry channel with a note, that the event will be fired within the next interaction with the expired file. This gives us a simplest-possible implementation of the "logical" Expiration of the uploads, in terms of Expired -> Not found (similar to what happens with the Terminated uploads). The "physical" expiration then might be whatever the developer would like to implement. It can be a simple cron job, that scans the uploads directory once per day, and removes all the files where expires_at < now(), or a super-smart system, keeping a full registry via pre-create and pre-finish hooks, or a goroutine, fired via the PostCreate callback, running alongside with the tusd server, if embedding as library, or anything else. This part we can leave to the devs using tusd. The protocol itself does not "force" us to clean up resources immediately, only to stop recognising the expired resource. 

What do you think overall about such solution?

**Comment by rglKali** [comment]
Hey everyone, any feedback on my PR is greatly appreciated!

**Comment by Acconut** [comment]
Thank you very much for taking this on, @rglKali, and thanks for opening a PR!

I agree with your assesment that a dedicated cleanup worker is overkill and outside of tusd's scope. In fact, I think that tusd could even leave the cleaning up entirely to the data storage backend and focus on restricting interations with uploads that have expired. As you proposed, tusd would check if an upload is expired when handling a new request, and refuse if that's the case. But the cleaning up is then the responsibility of the corresponding backend. For disk-based storage this could be a cron job finding files whose modification date exceeds a threshold. On S3, it could be an lifetime policy deleting files after 24hrs.

Any operator of tusd already needs to have such a mechansim in place anyways or otherwise unfinished uploads will stick around forever. In addition, uploads that aren't accessed past their expiration date wouldn't be cleaned otherwise. So instead of duplicating this logic in tusd, we can just take advantage of the existing cleanup logic. This would also avoid potential races between tusd and other jobs cleaning up files.

What do you think about that? That being said, your PR is a great starting point!
